### DISCUSSION TITLE
Altair Editor

### DISCUSSION THREAD

**Discussion by dwootton** [vega/altair#3470] [Ideas]
I would love the ability to share links to altair code + charts in a similar way to how we can with [the vega editor](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEzjQATjRyZ289AEEABNCTY46nILaZjMNgHcassobvqIcHA-OGks5YYINIJqhjg+FIJWshCGAGZs6obESJpsDLFQqRCUiiqmKKigDOqCaPmYSARZ7lIy8gC+SsjqANblXk6YeXCyUGzKTmRooAAeIyBxNHCCyuUAEokeXvbmeZgAnuHlAI4MfjqVOqQgTSAbE1Mzc+gAssFqAPrh6o8A4khRcutbcLv7skOphoJwaYKAA). 

This is particularly helpful when prototyping new visualizations as I can save the current version of a chart in a shareable way that others can remix. Could also be an easy way for people to get started writing altair code through an online editor that we can link to from the example galleries. 

Some initial options could be a pyodide based editor like [py cafe](https://py.cafe/) or a jupyterlite approach used by [nblink](https://github.com/dwootton/nblink). However, these approaches won't necessarily support altair-specific features likes showing signal or dataset values (re discussions with @jonmmease at SciPy). I find these feature instrumental when creating complex visualizations, and right now my workflow is to build the chart in the vega editor and then translate the syntax into altair after I'm finished.

**Comment by maartenbreddels** [comment]
Hi Dylan,

Altair works on https://py.cafe already, because we support [solara](https://github.com/widgetti/solara/), which can render anything Jupyter can, e.g.:
https://py.cafe/maartenbreddels/altair-car-performance-comparison

I wonder if an editor like you describe could be build using Solara/AnyWidget. We could also consider making a special Altair tile on our landing page (like we did with vizro), which would hide some of the wrapper code for setting up the editor.


What is also interesting to know, is that PyCafe can store a preview image:
[![preview](https://py.cafe/preview/maartenbreddels/altair-car-performance-comparison)](https://py.cafe/maartenbreddels/altair-car-performance-comparison)

These are visible from the user profile as well, e.g.: https://py.cafe/maartenbreddels/
This can be useful for galleries/examples pages.

We are also discussing with the Vizro developers to integrate into their docs. For instance, having an embedded code editor and embedded output in the documenation.
The embedded output is already supported (Share->Embed), and should give you code like:
```
<iframe src="https://py.cafe/embed/maartenbreddels/altair-car-performance-comparison?theme=light&linkToApp=false" width="100%" height="400px" style="border: 1px solid #e6e6e6; border-radius: 8px;"></iframe>
```

Let us know what you think, we are open to suggestions and feedback.

Regards,

Maarten

**Comment by jonmmease** [comment]
I wonder if this would make sense as a standalone PyScript app. See Altair example at https://pyscript.com/@examples/altair/latest. PyScript can interact with the dom directly (like in https://pyscript.com/@examples/d3-visualization/latest), so it should be possible to access the Vega view from Python in order to implement something like the data and signal explorers in the Vega editor.

**Comment by joelostblom** [comment]
Exciting idea! I would love to see something like the Vega editor for altair code, I think that would be useful also for the docs so that learners can interact with the code directly if there is something they want to try out. Panel's has set up their docs like that via pyodide and I tried to get it started previously in https://github.com/vega/altair/pull/2593, but the approaches here sound more promising, particularly the ability to inspect the transformed data.

**Comment by maartenbreddels** [comment]
Indeed, Jon, PyCafe does not allow you to interact directly with the DOM. It behaves nearly identically to a Python app, as if it did not run in pyodide/browser, including a normal `requirements.txt` file.

To access the vega signals, you'd have to follow the usual route if you want to build this for Jupyter or Solara. This means adding support for signals in the JupyterChart widget. But then, it would run on any platform (Jupyter, Solara, Google Colab, etc.) including PyCafe.

I did a simple proof of concept Altair Editor at:
https://py.cafe/maartenbreddels/altair-editor-poc

(PS: I just added a signals trait to the JupyterChart, and render only the signals using ipydatagrid - only now do I see there is some feature to read off the signals in the existing widget.)

**Reply by jonmmease** [reply]
Thanks for the POC @maartenbreddels, doing this using JupyterChart is an interesting idea!

**Comment by dangotbanned** [comment]
Still a work-in-progress, but we're using https://github.com/pyscript/pyscript to interact with the DOM from the docs in https://github.com/vega/altair/pull/3630#issuecomment-2411420348

https://github.com/user-attachments/assets/c8ef5032-7368-4fe9-b0c0-b48da2410f2c

It is a much simpler use-case, but may be helpful if anyone pursues that route
